### ISSUE TITLE
Why didn't you normalize each sub-feature?

### ISSUE THREAD

**Issue by SchafferZhang** [KikimorMay/MultiFace#2]
Hi, 
Thank you for your excellent work. I have a question about the implementation details in line 583 in model.py:
cos_theta = torch.mm(embbedings[:, i * self.each_embeding_size:(i + 1) * self.each_embeding_size],
                                 kernel_norm)
Why didn't you normalize the sub-feature " l2_norm(i * self.each_embeding_size:(i + 1) * self.each_embeding_size)",since the cosface needs the input and the weights to be normalized simultaneously?
Looking forward to your relay!

**Comment by KikimorMay** [comment]
Yes, we have tried to normalize the sub-features in line195 with model.py (the output embedding feature of the backbone), but the performance gets worse. We guess that the strong constraint on sub-features prevents the performance of the models, so we remove it and just normalize the weights in head kernels.  In our implementation, the norm of each sub-feature is slightly different but the accuracies get higher. 
Thank you for asking that !

**Comment by marigoold** [comment]
> Yes, we have tried to normalize the sub-features in line195 with model.py (the output embedding feature of the backbone), but the performance gets worse. We guess that the strong constraint on sub-features prevents the performance of the models, so we remove it and just normalize the weights in head kernels. In our implementation, the norm of each sub-feature is slightly different but the accuracies get higher.
> Thank you for asking that !

And you didn't normalize the original feature(dim=embedding_size)?

**Comment by KikimorMay** [comment]
> > Yes, we have tried to normalize the sub-features in line195 with model.py (the output embedding feature of the backbone), but the performance gets worse. We guess that the strong constraint on sub-features prevents the performance of the models, so we remove it and just normalize the weights in head kernels. In our implementation, the norm of each sub-feature is slightly different but the accuracies get higher.
> > Thank you for asking that !
> 
> And you didn't normalize the original feature(dim=embedding_size)?

We have normalized the original feature in line195 in model.py.
